x/retry/doc.go

```diff
@@ -0,0 +1,122 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package retry provides a YARPC middleware which is able to retry failed
+// outbound unary requests.
+//
+// Usage
+//
+// To build a retry middleware from config, first decode your configuration into
+// a `map[string]interface{}`
+//
+//  var data map[string]interface{}
+//  err := yaml.Unmarshal(myYAMLConfig, &data)
+//  mw, err := retry.NewOutboundMiddlewareFromConfig(data)
+//
+// Retry middleware can also be built by creating a PolicyProvider and passing
+// it in as an option.
+//
+//  mw := retry.NewOutboundMiddleware(retry.WithPolicyProvider(policyProvider))
+//
+// Check out the PolicyProvider docs for more details.
+//
+// Configuration
+//
+// The configuration may be specified in YAML or as any Go-level
+// map[string]interface{}. The examples below use YAML for illustration
+// purposes but other markup formats may be parsed into map[string]interface{}
+// as long as the information provided is the same.
+//
+// The configuration accepts the following top-level attributes: policies,
+// default, and overrides.
+//
+//  policies:
+//    # ...
+//  default: ...
+//  overrides:
+//    # ...
+//
+// See the following sections for details on the policies, default, and
+// overrides keys in the configuration.
+//
+// Policies Configuration
+//
+// The 'policies' attribute configures different retry policies which can be
+// referenced individually in the 'default' and 'override' sections.
+//
+//  policies:
+//    fastretry:
+//      retries: 5
+//      maxtimeout: 10ms
+//      backoff:
+//        exponential:
+//          first: 5ms
+//          max: 1s
+//    slowretry:
+//      retries: 3
+//      maxtimeout: 100ms
+//      backoff:
+//        exponential:
+//          first: 5ms
+//          max: 10s
+//
+// This configuration would create two retry policies we could use later.
+//
+// "fastretry" will max out request timeouts to 10 ms, and will re-attempt
+// requests 5 times before failing.  Between every failure, it will initially
+// use a full jitter exponential backoff of 5ms with a max backoff of 1s.
+//
+// "slowretry" will max out request timeouts to 100 ms, and will re-attempt
+// requests 3 times before failing.  Between every failure, it will initially
+// use a full jitter exponential backoff of 5ms with a max backoff of 10s.
+//
+// Default Configuration
+//
+// The 'default' attributes indicates which "policy" will be the default for
+// all retry attempts.
+//
+// 	default: slowretry
+//
+// 'slowretry' needs to be a defined policy in the 'policies' section (or it
+// needs to be one of the default policies we've provided).
+//
+// Overrides Configuration
+//
+// The 'overrides' attribute configures custom retry policies for specific
+// services or procedures.
+//
+//  overrides:
```

Comment:
Nit: I don't know if we've discussed the terminology "override" much. This is less override and more matching. You're specifying policies that apply to different matching requests and specifying a fallback ('default') for cases where there's no match. I can't think of an obvious alternative name right now but bringing this up in case you have another name in mind.